### PR TITLE
Fix (grainfmt): better handling of multiple line items and comments

### LINKED ISSUE #1444: Formatter: issues with annotated types
type Foo<a /* comment1 */,b /* comment2 */> = (List<a>, List<b>)

formats to

type Foo<a, /* comment1 */ b> = ( /* comment2 */
  List< /* comment1 */ /* comment2 */a>,
  List< /* comment1 */ /* comment2 */b>
)

### PR DESCRIPTION
**Pull request by marcusroberts**
fixes #1444  and other pieces of code with multiple items (e.g. record definitions) to handle items that break over multiple lines, comments that force line breaks, and positioning.

Apologies, the git diff is hard to parse - item_iterator is the big change.

Also some tests change as the implementation chooses a different arbitrary position for comments before/after comments as comments aren't recorded in the AST

Rewrote the code that iterates over code that has multiple items separated by e.g. commas, e.g. record definitions, lists of args to functions, and handles placing comments better.  The code is a lot simpler now too.

### CODE REVIEW

File: compiler/src/formatting/format.re [hunk 1]
```diff
@@ -726,239 +726,219 @@ let rec block_item_iterator =
...
+            );
           Doc.concat([
-            Comment_utils.new_comments_to_docs(leading_comments),
-            filler,
-            print_item(item, ~comments=item_comments),
+            leading_comments_docs,
+            itemdoc,
+            Doc.ifBreaks(
+              if (trailing_separator) {separator} else {Doc.nil},
+              // Doc.text("space"),
```
Comment (ospencer) [inline compiler/src/formatting/format.re:911]: Should this be deleted?

File: compiler/src/formatting/format.re [hunk 2]
```diff
@@ -3764,14 +3766,28 @@ let rec print_data =
...
+    // | [] =>
+    //   Comment_utils.get_after_brace_comments(~loc=data.pdata_loc, comments)
+    // | [first, ...rem] =>
+    //  Comment_utils.get_after_brace_comments(~loc=data.pdata_loc, comments)
+    //   // Comment_utils.get_comments_enclosed_and_before_location(
+    //   //   ~loc1=data.pdata_name.loc,
+    //   //   ~loc2=first.pld_loc,
+    //   //   comments,
+    //   // )
+    // };
```
Comment (ospencer) [inline compiler/src/formatting/format.re:3779]: What's up with this?

**Review by ospencer — COMMENTED** [review]
Can you explain a little more about what this PR is doing?

File: compiler/src/formatting/format.re (hunk 2 above)
Comment (marcusroberts) [reply compiler/src/formatting/format.re:3779]: I swore I deleted that!  Fixing

File: compiler/src/formatting/format.re [hunk 3]
```diff
@@ -3505,80 +3494,89 @@ let rec print_data =
...
+        };
 
       let items =
         item_iterator(
           ~get_loc,
           ~print_item,
-          ~comments=cleaned_comments,
+          ~comments=type_comments,
           ~separator=Doc.comma,
+          ~trailing_separator=false,
```
Comment (phated) [inline compiler/src/formatting/format.re:3542]: Refactor to do away with boolean param

File: compiler/src/formatting/format.re [hunk 4]
```diff
@@ -3621,6 +3619,7 @@ let rec print_data =
                   ~print_item,
                   ~comments=cleaned_comments,
                   ~separator=Doc.comma,
+                  ~trailing_separator=false,
```
Comment (phated) [inline compiler/src/formatting/format.re:3622]: Refactor to do away with boolean param

File: compiler/src/formatting/format.re [hunk 5]
```diff
@@ -3661,6 +3660,7 @@ let rec print_data =
         ~print_item,
         ~comments=cleaned_comments,
         ~separator=Doc.comma,
+        ~trailing_separator=true,
```
Comment (phated) [inline compiler/src/formatting/format.re:3663]: Refactor to do away with boolean param

File: compiler/src/formatting/format.re [hunk 6]
```diff
@@ -3704,6 +3704,7 @@ let rec print_data =
               ~print_item,
               ~comments=cleaned_comments,
               ~separator=Doc.comma,
+              ~trailing_separator=false,
```
Comment (phated) [inline compiler/src/formatting/format.re:3707]: Refactor to do away with boolean param

File: compiler/src/formatting/format.re [hunk 7]
```diff
@@ -3736,7 +3737,7 @@ let rec print_data =
         Doc.lbrace,
         Comment_utils.single_line_of_comments(after_brace_comments),
         Doc.indent(printed_decls_after_brace),
-        Doc.ifBreaks(Doc.comma, Doc.nil),
+        //  Doc.ifBreaks(Doc.comma, Doc.nil),
```
Comment (phated) [inline compiler/src/formatting/format.re:3740]:
```suggestion
```
Delete if not used?

File: compiler/src/formatting/format.re [hunk 8]
```diff
@@ -3764,14 +3765,17 @@ let rec print_data =
...
     let cleaned_comments =
       remove_used_comments(~remove_comments=after_brace_comments, comments);
+
     let decl_items =
       item_iterator(
         ~get_loc,
         ~print_item,
         ~comments=cleaned_comments,
         ~separator=Doc.comma,
+        ~trailing_separator=true,
```
Comment (phated) [inline compiler/src/formatting/format.re:3778]: Refactor to do away with boolean param

File: compiler/src/formatting/format.re [hunk 9]
```diff
@@ -3807,6 +3811,7 @@ let rec print_data =
               ~print_item,
               ~comments=cleaned_comments,
               ~separator=Doc.comma,
+              ~trailing_separator=false,
```
Comment (phated) [inline compiler/src/formatting/format.re:3814]: Refactor to do away with boolean param

File: compiler/test/formatter_outputs/comments.gr [hunk 10]
```diff
@@ -3,15 +3,15 @@
  followed by a blank line
 */
 
-let myfun = (/* a */ x, /*post*/ y) => 10
+let myfun = (/* a */ x /*post*/, y) => 10
```
Comment (phated) [inline compiler/test/formatter_outputs/comments.gr:6]: Why did you change the output locations of this?

File: compiler/test/formatter_outputs/comments.gr [hunk 11]
```diff
@@ -3,15 +3,15 @@
...
 
 /**
  * Then a doc
  */
 // Then a line
 let myfun1 = (/* a */ x, y) => 10
 
-let myfun2 = (x, /*post*/ y) => 10
+let myfun2 = (x /*post*/, y) => 10
```
Comment (phated) [inline compiler/test/formatter_outputs/comments.gr:14]: Why did you change the output locations of this?

File: compiler/test/formatter_outputs/comments.gr [hunk 12]
```diff
@@ -77,7 +77,7 @@ if (a /*why*/) {
   // end block
 }
 
-let myfun /*special*/ = (/*lead*/ x, /*follow*/ y) => x + 5
+let myfun /*special*/ = (/*lead*/ x /*follow*/, y) => x + 5
```
Comment (phated) [inline compiler/test/formatter_outputs/comments.gr:80]: Why did you change the output locations of this?

**Review by phated — CHANGES_REQUESTED** [review]
Some thoughts and comments. One thing that I can't help but wonder is if the abstractions are wrong (because you always need to pass extra flags when you try to customize behavior)... though that is probably a bigger discussion than this PR.

File: compiler/test/formatter_outputs/comments.gr (hunk 12 above)
Comment (marcusroberts) [reply compiler/test/formatter_outputs/comments.gr:80]: Because the comma isn't in the AST, the choice of where to place it in relation to the comments is arbitrary, and the new way of processing chooses the other.

File: compiler/test/formatter_outputs/comments.gr (hunk 11 above)
Comment (marcusroberts) [reply compiler/test/formatter_outputs/comments.gr:14]: Because the comma isn't in the AST, the choice of where to place it in relation to the comments is arbitrary, and the new way of processing chooses the other.

File: compiler/test/formatter_outputs/comments.gr (hunk 10 above)
Comment (marcusroberts) [reply compiler/test/formatter_outputs/comments.gr:6]: (Same as others, decision is arbitrary, see below)

File: compiler/src/formatting/format.re [hunk 13]
```diff
@@ -726,239 +726,218 @@ let rec block_item_iterator =
...
-          comments,
+let rec item_iterator =
+        (
+          ~previous: option(Location.t)=?,
+          ~get_loc: 'a => Location.t,
+          ~print_item: (~comments: list(Parsetree.comment), 'a) => Doc.t,
+          ~comments: list(Parsetree.comment),
+          ~separator,
+          ~followed_by_arrow: option(bool)=?,
+          ~trailing_separator: bool,
```
[the remark this answers is not in the document]
Comment (marcusroberts) [reply compiler/src/formatting/format.re:847]: I believe this was refactored away in a later PR, let me check

[the remark this answers is not in the document]
Comment (marcusroberts) [reply compiler/src/formatting/format.re:847]: This is a boolean.  It's a style preference, do we want to add a trailing separator to a set of items if it breaks over multiple lines.  For example lists do, records don't.

File: compiler/src/formatting/format.re [hunk 14]
```diff
@@ -726,239 +726,218 @@ let rec block_item_iterator =
...
+              ~location=get_loc(item),
+              comments,
+            );
+          switch (leading_comments) {
+          | [] => Doc.nil
+          | _ =>
+            Doc.group(
+              Doc.concat([
+                Comment_utils.new_comments_to_docs(leading_comments),
+                Doc.ifBreaks(Doc.nil, Doc.space),
```
[the remark this answers is not in the document]
Comment (marcusroberts) [reply compiler/src/formatting/format.re:866]: You're thinking of softLine.  But that acts a little differently.

File: compiler/src/formatting/format.re (hunk 13 above)
[the remark this answers is not in the document]
Comment (ospencer) [reply compiler/src/formatting/format.re:847]: We should do that for records though.

File: compiler/src/formatting/format.re [hunk 15]
```diff
@@ -726,239 +726,218 @@ let rec block_item_iterator =
...
+            Doc.ifBreaks(
+              if (trailing_separator) {separator} else {Doc.nil},
+              Doc.nil,
+            ),
+            trailing_cmts,
           ]);
+        };
+      } else {
+        // look for comments between the current item and the next item
+        // we know we have one more item
```
[the remark this answers is not in the document]
Comment (marcusroberts) [reply compiler/src/formatting/format.re:918]: Because we are in an the else of an if statement that tested for being the last item we are mapping over

File: compiler/src/formatting/format.re [hunk 16]
```diff
@@ -998,6 +977,7 @@ let rec resugar_list_patterns =
       ~print_item,
       ~comments,
       ~separator=Doc.comma,
+      ~trailing_separator=true,
```
[the remark this answers is not in the document]
Comment (marcusroberts) [reply compiler/src/formatting/format.re:980]: Can only do this if every iterator requires a trailing separator.

File: compiler/test/formatter_outputs/comments.gr (hunk 12 above)
Comment (phated) [reply compiler/test/formatter_outputs/comments.gr:80]: Why does it choose that? I prefer the old way.

File: compiler/src/formatting/format.re [hunk 17]
```diff
@@ -1301,6 +1281,7 @@ and print_record_pattern =
       ~print_item,
       ~comments=cleaned_comments,
       ~separator=Doc.comma,
+      ~trailing_separator=false,
```
[the remark this answers is not in the document]
Comment (marcusroberts) [reply compiler/src/formatting/format.re:1284]: Can only do this if every iterator requires a trailing separator. e.g. enums

File: compiler/src/formatting/format.re [hunk 18]
```diff
@@ -726,239 +726,218 @@ let rec block_item_iterator =
...
+            Doc.group(
+              Doc.concat([
+                Comment_utils.new_comments_to_docs(leading_comments),
+                Doc.ifBreaks(Doc.nil, Doc.space),
+              ]),
+            )
+          };
+        } else {
+          Doc.nil;
+        };
```
[the remark this answers is not in the document]
Comment (marcusroberts) [reply compiler/src/formatting/format.re:872]: Ignore this, I rewrote it in half the lines :-D

File: compiler/src/formatting/format.re (hunk 16 above)
[the remark this answers is not in the document]
Comment (marcusroberts) [reply compiler/src/formatting/format.re:980]: Refactored to do away with the boolean param

File: compiler/src/formatting/format.re (hunk 17 above)
[the remark this answers is not in the document]
Comment (marcusroberts) [reply compiler/src/formatting/format.re:1284]: Refactored so this decision is made in one place

File: compiler/src/formatting/format.re (hunk 13 above)
[the remark this answers is not in the document]
Comment (marcusroberts) [reply compiler/src/formatting/format.re]: The decision making for trailing separators is in one place now.  We should review what should have them and update as needed

File: compiler/src/formatting/format.re [hunk 19]
```diff
@@ -726,239 +726,218 @@ let rec block_item_iterator =
...
-        | in_comments =>
-          let last_comment = get_last_item_in_list(in_comments);
+let print_trailing_comments = (~separator, ~itemloc: Location.t, comments) => {
+  let previous = ref(None);
 
-          switch (last_comment) {
-          | Line(_) => Doc.breakParent
-          | _ => Doc.nil
+  let items =
+    List.fold_left(
```
[the remark this answers is not in the document]
Comment (marcusroberts) [reply compiler/src/formatting/format.re:767]: This is tricky.   I'm working top to bottom, left to right, so the fold_left works here.   The lists aren't long.

Trying to work from right to left is much harder

File: compiler/src/formatting/format.re (hunk 15 above)
[the remark this answers is not in the document]
Comment (marcusroberts) [reply compiler/src/formatting/format.re]: Code no longer in use

File: compiler/test/formatter_outputs/comments.gr (hunk 12 above)
Comment (marcusroberts) [reply compiler/test/formatter_outputs/comments.gr:80]: Before I was looking for leading comments, but that broke how mixed comments and separators where we need to consider trailing comments.

This way, this formats better:  50% of one, half the other IMO

`let myfun /*special*/ = (x /*lead*/, y /*follow*/) => x + 5`

**Comment by marcusroberts** [comment]
Merged in master so this can be merged safely.

File: compiler/src/formatting/format.re [hunk 20]
```diff
@@ -6,6 +6,22 @@ open Grain_diagnostics;
...
+  | Iterated_Match_Item
+  | Iterated_Data_Declarations
+  | Iterated_Record_Labels
+  | Iterated_Tuple_Expression
+  | Iterated_Array_Expression
+  | Iterated_Type_Items
+  | Iterated_Tuple_Constructor
+  | Iterated_Enum
+  | Iterated_Record_Data
+  | Iterated_Value_Bindings;
```
Comment (phated) [inline compiler/src/formatting/format.re:24]:
```suggestion
  | IteratedListPattern
  | IteratedRecordPattern
  | IteratedRecord
  | IteratedTypeConstructor
  | IteratedPatterns
  | IteratedMatchItem
  | IteratedDataDeclarations
  | IteratedRecordLabels
  | IteratedTupleExpression
  | IteratedArrayExpression
  | IteratedTypeItems
  | IteratedTupleConstructor
  | IteratedEnum
  | IteratedRecordData
  | IteratedValueBindings;
```
In this file, we've been using snake-case for the type names (because ocaml needs them lowercase) and variants are PascalCase

File: compiler/src/formatting/format.re [hunk 21]
```diff
@@ -744,238 +760,230 @@ let rec block_item_iterator =
...
+    | Iterated_Record
+    | Iterated_Type_Constructor
+    | Iterated_Patterns
+    | Iterated_Tuple_Expression
+    | Iterated_Array_Expression
+    | Iterated_Type_Items
+    | Iterated_Tuple_Constructor
+    | Iterated_Enum
+    | Iterated_Record_Data
+    | Iterated_Value_Bindings => false
```
Comment (phated) [inline compiler/src/formatting/format.re:898]:
```suggestion
    | IteratedListPattern
    | IteratedDataDeclarations
    | IteratedMatchItem
    | IteratedRecordLabels => true
    | IteratedRecordPattern
    | IteratedRecord
    | IteratedTypeConstructor
    | IteratedPatterns
    | IteratedTupleExpression
    | IteratedArrayExpression
    | IteratedTypeItems
    | IteratedTupleConstructor
    | IteratedEnum
    | IteratedRecordData
    | IteratedValueBindings => false
```

File: compiler/src/formatting/format.re [hunk 22]
```diff
@@ -1016,6 +1024,7 @@ let rec resugar_list_patterns =
       ~print_item,
       ~comments,
       ~separator=Doc.comma,
+      ~iterated_item=Iterated_List_Pattern,
```
Comment (phated) [inline compiler/src/formatting/format.re:1027]:
```suggestion
      ~iterated_item=IteratedListPattern,
```

File: compiler/src/formatting/format.re [hunk 23]
```diff
@@ -1321,6 +1330,7 @@ and print_record_pattern =
       ~print_item,
       ~comments=cleaned_comments,
       ~separator=Doc.comma,
+      ~iterated_item=Iterated_Record_Pattern,
```
Comment (phated) [inline compiler/src/formatting/format.re:1333]:
```suggestion
      ~iterated_item=IteratedRecordPattern,
```

File: compiler/src/formatting/format.re [hunk 24]
```diff
@@ -1606,6 +1616,7 @@ and print_record =
       ~print_item,
       ~comments=cleaned_comments,
       ~separator=Doc.comma,
+      ~iterated_item=Iterated_Record,
```
Comment (phated) [inline compiler/src/formatting/format.re:1619]:
```suggestion
      ~iterated_item=IteratedRecord,
```

File: compiler/src/formatting/format.re [hunk 25]
```diff
@@ -1729,6 +1740,7 @@ and print_type =
           ~print_item,
           ~comments=cleaned_comments,
           ~separator=Doc.comma,
+          ~iterated_item=Iterated_Type_Constructor,
```
Comment (phated) [inline compiler/src/formatting/format.re:1743]:
```suggestion
          ~iterated_item=IteratedTypeConstructor,
```

File: compiler/src/formatting/format.re [hunk 26]
```diff
@@ -2450,6 +2461,7 @@ and print_patterns =
         ~comments=comments_in_scope,
         ~separator=Doc.comma,
         ~followed_by_arrow?,
+        ~iterated_item=Iterated_Patterns,
```
Comment (phated) [inline compiler/src/formatting/format.re:2464]:
```suggestion
        ~iterated_item=IteratedPatterns,
```

File: compiler/src/formatting/format.re [hunk 27]
```diff
@@ -2541,6 +2553,7 @@ and print_expression =
           ~print_item,
           ~comments=cleaned_comments,
           ~separator=Doc.comma,
+          ~iterated_item=Iterated_Tuple_Expression,
```
Comment (phated) [inline compiler/src/formatting/format.re:2556]:
```suggestion
          ~iterated_item=IteratedTupleExpression,
```

File: compiler/src/formatting/format.re [hunk 28]
```diff
@@ -2590,6 +2603,7 @@ and print_expression =
           ~print_item,
           ~comments=cleaned_comments,
           ~separator=Doc.comma,
+          ~iterated_item=Iterated_Array_Expression,
```
Comment (phated) [inline compiler/src/formatting/format.re:2606]:
```suggestion
          ~iterated_item=IteratedArrayExpression,
```

File: compiler/src/formatting/format.re [hunk 29]
```diff
@@ -2847,6 +2861,7 @@ and print_expression =
           ~print_item,
           ~comments=cleaned_comments,
           ~separator=Doc.comma,
+          ~iterated_item=Iterated_Match_Item,
```
Comment (phated) [inline compiler/src/formatting/format.re:2864]:
```suggestion
          ~iterated_item=IteratedMatchItem,
```

File: compiler/src/formatting/format.re [hunk 30]
```diff
@@ -3738,6 +3752,7 @@ and print_value_bind =
         ~print_item,
         ~comments,
         ~separator=Doc.comma,
+        ~iterated_item=Iterated_Value_Bindings,
```
Comment (phated) [inline compiler/src/formatting/format.re:3755]:
```suggestion
        ~iterated_item=IteratedValueBindings,
```

File: compiler/src/formatting/format.re [hunk 31]
```diff
@@ -3820,80 +3838,89 @@ let rec print_data =
...
+        };
 
       let items =
         item_iterator(
           ~get_loc,
           ~print_item,
-          ~comments=cleaned_comments,
+          ~comments=type_comments,
           ~separator=Doc.comma,
+          ~iterated_item=Iterated_Type_Items,
```
Comment (phated) [inline compiler/src/formatting/format.re:3886]:
```suggestion
          ~iterated_item=IteratedTypeItems,
```

File: compiler/src/formatting/format.re [hunk 32]
```diff
@@ -3936,6 +3963,7 @@ let rec print_data =
                   ~print_item,
                   ~comments=cleaned_comments,
                   ~separator=Doc.comma,
+                  ~iterated_item=Iterated_Tuple_Constructor,
```
Comment (phated) [inline compiler/src/formatting/format.re:3966]:
```suggestion
                  ~iterated_item=IteratedTupleConstructor,
```

File: compiler/src/formatting/format.re [hunk 33]
```diff
@@ -3976,6 +4004,7 @@ let rec print_data =
         ~print_item,
         ~comments=cleaned_comments,
         ~separator=Doc.comma,
+        ~iterated_item=Iterated_Data_Declarations,
```
Comment (phated) [inline compiler/src/formatting/format.re:4007]:
```suggestion
        ~iterated_item=IteratedDataDeclarations,
```

File: compiler/src/formatting/format.re [hunk 34]
```diff
@@ -4019,6 +4048,7 @@ let rec print_data =
               ~print_item,
               ~comments=cleaned_comments,
               ~separator=Doc.comma,
+              ~iterated_item=Iterated_Enum,
```
Comment (phated) [inline compiler/src/formatting/format.re:4051]:
```suggestion
              ~iterated_item=IteratedEnum,
```

File: compiler/src/formatting/format.re [hunk 35]
```diff
@@ -4079,14 +4108,17 @@ let rec print_data =
...
     let cleaned_comments =
       remove_used_comments(~remove_comments=after_brace_comments, comments);
+
     let decl_items =
       item_iterator(
         ~get_loc,
         ~print_item,
         ~comments=cleaned_comments,
         ~separator=Doc.comma,
+        ~iterated_item=Iterated_Record_Labels,
```
Comment (phated) [inline compiler/src/formatting/format.re:4121]:
```suggestion
        ~iterated_item=IteratedRecordLabels,
```

File: compiler/src/formatting/format.re [hunk 36]
```diff
@@ -4122,6 +4154,7 @@ let rec print_data =
               ~print_item,
               ~comments=cleaned_comments,
               ~separator=Doc.comma,
+              ~iterated_item=Iterated_Record_Data,
```
Comment (phated) [inline compiler/src/formatting/format.re:4157]:
```suggestion
              ~iterated_item=IteratedRecordData,
```

File: compiler/test/formatter_outputs/comments.gr (hunk 10 above)
Comment (phated) [reply compiler/test/formatter_outputs/comments.gr:6]: Ah, so it's actually *better* at keeping the input location now. Nice!

File: compiler/test/formatter_outputs/comments.gr (hunk 11 above)
Comment (phated) [reply compiler/test/formatter_outputs/comments.gr:16]: Ah, so it's actually *better* at keeping the input location now. Nice!

File: compiler/test/formatter_outputs/comments.gr (hunk 12 above)
Comment (phated) [reply compiler/test/formatter_outputs/comments.gr:82]: Ah, so it's actually *better* at keeping the input location now. Nice!

File: package-lock.json [hunk 37]
```diff
@@ -19,11 +19,11 @@
     },
     "cli": {
       "name": "@grain/cli",
-      "version": "0.4.7",
```
Comment (phated) [inline package-lock.json:22]: These changes seem wrong (we don't update them in the lock 🤷 )

**Review by phated — CHANGES_REQUESTED** [review]
Some consistency suggestions. I also want to try poking at the `List.append` stuff a bit

File: compiler/src/formatting/format.re (hunk 20 above)
Comment (marcusroberts) [reply compiler/src/formatting/format.re:24]: Sorry, agreed!  I've fixed them in one go locally rather than accept your commit suggestions so I don't trigger a thousand CI jobs.

Comment (marcusroberts) [reply compiler/src/formatting/format.re]: Found how to batch it :)

File: compiler/src/formatting/format.re [hunk 38]
```diff
@@ -744,238 +760,252 @@ let rec block_item_iterator =
...
+    | IteratedTypeConstructor
+    | IteratedPatterns
+    | IteratedTupleExpression
+    | IteratedArrayExpression
+    | IteratedTypeItems
+    | IteratedTupleConstructor
+    | IteratedEnum
+    | IteratedRecordData
+    | IteratedValueBindings => false
+    };
```
Comment (phated) [inline compiler/src/formatting/format.re:921]: @ospencer to review

File: compiler/src/formatting/format.re [hunk 39]
```diff
@@ -744,238 +760,252 @@ let rec block_item_iterator =
...
-                  ),
-                ]),
-              )
-            },
-          ]);
+  let number_items = List.length(items);
+
+  if (number_items < 1) {
+    []; // nothing to do
+  } else {
```
Comment (phated) [inline compiler/src/formatting/format.re:927]: Can we turn this into a pattern match?

```
switch (items) {
  | [] => []
  | [first_item, ...rest] => { ... }
}
```
avoids the `List.length` and `List.hd` calls.

**Review by phated — CHANGES_REQUESTED** [review]
One more opportunity for code cleanup. Otherwise looks good to me. I also want Oscar to look over the trailing_separator definiton

File: compiler/src/formatting/format.re (hunk 39 above)
Comment (marcusroberts) [reply compiler/src/formatting/format.re:927]: Of course. I seem to have a blind spot for these!

Comment (marcusroberts) [reply compiler/src/formatting/format.re:927]: We need the list length lower though.

**Review by phated — APPROVED** [review]
This looks great! Thanks!

File: compiler/src/formatting/format.re [hunk 40]
```diff
@@ -744,238 +760,251 @@ let rec block_item_iterator =
...
+          ~iterated_item: iterator_item_type,
+          items: list('a),
+        ) => {
+  let trailing_separator =
+    switch (iterated_item) {
+    | IteratedListPattern
+    | IteratedDataDeclarations
+    | IteratedMatchItem
+    | IteratedRecordLabels => true
+    | IteratedRecordPattern
```
Comment (ospencer) [inline compiler/src/formatting/format.re:909]: I would say this should be yes

File: compiler/src/formatting/format.re [hunk 41]
```diff
@@ -744,238 +760,251 @@ let rec block_item_iterator =
...
+    | IteratedRecord
+    | IteratedTypeConstructor
+    | IteratedPatterns
```
Comment (ospencer) [inline compiler/src/formatting/format.re:912]: when is this?

File: compiler/src/formatting/format.re [hunk 42]
```diff
@@ -744,238 +760,251 @@ let rec block_item_iterator =
...
+    | IteratedTupleExpression
+    | IteratedArrayExpression
```
Comment (ospencer) [inline compiler/src/formatting/format.re:914]: and this should be yes

File: compiler/test/formatter_outputs/aliases.gr [hunk 43]
```diff
@@ -23,15 +23,15 @@ type AReallyReallyReallyReallyReallyReallyReallyReallyReallyLongLineBreakingType
...
 > = String
 
-type Foo<
+type Foo3<
   a, // Test comment
   b
 > = String
 
-type Foo<a, /* Test comment */ b> = String
+type Foo4<a /* Test comment */, b> = String
```
Comment (ospencer) [inline compiler/test/formatter_outputs/aliases.gr:31]: this shouldn't have a trailing comma 🤔

File: compiler/src/formatting/format.re (hunk 42 above)
Comment (marcusroberts) [reply compiler/src/formatting/format.re:914]: By yes you mean trailing separators when over multiple lines?

Comment (ospencer) [reply compiler/src/formatting/format.re:914]: Right.

File: compiler/test/formatter_outputs/aliases.gr (hunk 43 above)
Comment (marcusroberts) [reply compiler/test/formatter_outputs/aliases.gr:31]: It doesn't, that would be 
`type Foo4<a /* Test comment */, b, > = String`

File: compiler/src/formatting/format.re (hunk 41 above)
Comment (marcusroberts) [reply compiler/src/formatting/format.re:912]: Sorry, can you give more detail?

File: compiler/test/formatter_outputs/aliases.gr (hunk 43 above)
Comment (ospencer) [reply compiler/test/formatter_outputs/aliases.gr:31]: my eyes haha

File: compiler/test/formatter_outputs/aliases.gr [hunk 44]
```diff
@@ -6,32 +6,32 @@ type Foo<a> = String
...
 type Foo<a, b> = String
 
-type Foo<a, b> = String // Test comment
+type Foo1<a, b> = String // Test comment
 
-type Foo< // Test comment
+type Foo2< // Test comment
   a,
-  b
+  b,
```
Comment (ospencer) [inline compiler/test/formatter_outputs/aliases.gr:13]: I might not have understood what was what, but this shouldn't be trailing.

File: compiler/test/stdlib/buffer.test.gr [hunk 45]
```diff
@@ -207,7 +207,7 @@ let slices = [>
   ("0123456789", -1, -1),
   ("0123456789", 2, 3),
   ("0123456789", 0, 10),
-  ("0123456789", 6, 6),
+  ("0123456789", 6, 6),,
```
Comment (ospencer) [inline compiler/test/stdlib/buffer.test.gr:210]: this has two trailing commas

File: compiler/src/formatting/format.re (hunk 41 above)
Comment (ospencer) [reply compiler/src/formatting/format.re]: Just asking what type of code that references!

File: compiler/test/stdlib/buffer.test.gr (hunk 45 above)
Comment (marcusroberts) [reply compiler/test/stdlib/buffer.test.gr]: Yes, bad push, those tests are cleaned up again now

File: compiler/test/formatter_outputs/aliases.gr (hunk 44 above)
Comment (marcusroberts) [reply compiler/test/formatter_outputs/aliases.gr]: Reverted

File: compiler/test/formatter_outputs/application.gr [hunk 46]
```diff
@@ -81,5 +81,5 @@ let preExistingObjectsWithRefCountMismatch2 = Map.make(): Map.Map<
     Number,
     Number,
     Number
-  )
+  ),
```
Comment (ospencer) [inline compiler/test/formatter_outputs/application.gr:84]: Yeah, this is where I don't want a trailing comma.

Comment (marcusroberts) [reply compiler/test/formatter_outputs/application.gr:84]: Glad to hear that, I wondered why you wanted it, it looks ugly

File: compiler/test/formatter_outputs/enum_long.gr [hunk 47]
```diff
@@ -26,7 +26,7 @@ export enum Shape<
   z,
   aa,
   bb,
-  cc
+  cc,
```
Comment (ospencer) [inline compiler/test/formatter_outputs/enum_long.gr:29]: Don't want this trailing comma!

Comment (marcusroberts) [reply compiler/test/formatter_outputs/enum_long.gr]: reverted

**Review by ospencer — APPROVED** [review]
